helm: Support VolumeSnapshotClass and VolumeGroupSnapshotClass

hswong3i opened the pull request:
<!-- Please take a look at our [Contributing](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#Code-contribution-workflow)
documentation before submitting a Pull Request!
Thank you for contributing to ceph-csi! -->

# Describe what this PR does #

This PR add support for creating VolumeSnapshotClass and VolumeGroupSnapshotClass with helm charts. Both default with `create: false` as StorageClass support.

Moreover, it handle the chart release with branch `gh-pages`, by integrate with https://github.com/helm/chart-releaser-action.

The sample helm repo could find from https://alvistack.github.io/ceph-ceph-csi/index.yaml, by:

    helm repo add ceph-csi https://alvistack.github.io/ceph-ceph-csi
    helm repo update
    helm search repo ceph-csi

## Future concerns ##

List items that are not part of the PR and do not impact it's
functionality, but are work items that can be taken up subsequently.

**Checklist:**

* [X] **Commit Message Formatting**: Commit titles and messages follow
  guidelines in the [developer guide](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages).
* [x] Reviewed the developer guide on [Submitting a Pull Request](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#development-workflow)
* [x] [Pending release notes](https://github.com/ceph/ceph-csi/blob/devel/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next major release.
* [x] Documentation has been updated, if necessary.
* [ ] Unit tests have been added, if necessary.
* [ ] Integration tests have been added, if necessary.

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

* `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

nixpanic commented:
Hi @hswong3i, thanks for the PR!

Could you remove the publishing of the charts from your commit? There is a job already, and it pushes the charts to the https://github.com/ceph/csi-charts repository.

hswong3i commented:
> Hi @hswong3i, thanks for the PR!
> 
> Could you remove the publishing of the charts from your commit? There is a job already, and it pushes the charts to the https://github.com/ceph/csi-charts repository.

@nixpanic dummy https://github.com/helm/chart-releaser-action integration removed, also update `charts/*/README.md` accordingly with `pandoc -f gfm -t gfm --wrap preserve`.

nixpanic commented:
Thanks for the update @hswong3i! I've glanced over it just now, and this looks better than the previous version. It needs a more careful look to check the parameters/values before I feel confident enough to approve it.

iPraveenParihar commented on charts/ceph-csi-cephfs/README.md:
```diff
@@ -87,13 +87,13 @@ permissions will be granted globally via a *ClusterRole*.
 #### Known Issues Upgrading
 
-- When upgrading to version >=3.7.0, you might encounter an error that the
+- When upgrading to version \>=3.7.0, you might encounter an error that the
```
Is this a typo?

hswong3i replied:
I am cleaning up the doc with `pandoc -f gfm -t gfm --wrap preserve`, so the `>` and `<` get escaped due to not HTML safe.

I could manually revert these minor changes.

iPraveenParihar commented on charts/ceph-csi-cephfs/templates/groupsnapshotclass.yaml:
```diff
@@ -0,0 +1,30 @@
+{{- if .Values.volumeGroupSnapshotClass.groupSnapshotterSecretNamespace }}
+  csi.storage.k8s.io/group-snapshotter-secret-namespace: {{ .Values.volumeGroupSnapshotClass.groupSnapshotterSecretNamespace }}
+{{ else }}
+csi.storage.k8s.io/group-snapshotter-secret-namespace: {{ .Release.Namespace }}
```
why is this duplicate required?

hswong3i replied:
Similar logic are copy-and-pasted from https://github.com/ceph/ceph-csi/blob/cc407d157ec33e69d8b33652d5232c9187569624/charts/ceph-csi-cephfs/templates/storageclass.yaml#L42-L46

iPraveenParihar commented on charts/ceph-csi-cephfs/README.md:
```diff
@@ -118,118 +118,134 @@ kubectl delete namespace ceph-csi-cephfs
+| `volumeSnapshotClass.create`                               | Specifies whether the VolumeSnapshotClass should be created                                                                                                                                                                                                                   | `false`                                                 |
+| `volumeSnapshotClass.name`                                 | Specifies the cephFS VolumeSnapshotClass name                                                                                                                                                                                                                                 | `csi-cephfsplugin-snapclass`                            |
+| `volumeSnapshotClass.annotations`                          | Specifies the annotations for the cephFS volumeSnapshotClass                                                                                                                                                                                                                  | `[]`                                                    |
+| `volumeSnapshotClass.clusterID`                            | String representing a Ceph cluster to provision storage from                                                                                                                                                                                                                  | `<cluster-ID>`                                          |
```
```suggestion
| `volumeSnapshotClass.clusterID`                            | String representing a Ceph cluster to provision storage snapshot from                                                                                                                                                                                                                  | `<cluster-ID>`                                          |
```

iPraveenParihar replied:
apply to all places where applicable

iPraveenParihar commented on charts/ceph-csi-rbd/README.md:
```diff
@@ -91,143 +91,159 @@ kubectl delete namespace ceph-csi-rbd
+| `volumeSnapshotClass.name`                                 | Specifies the rbd VolumeSnapshotClass name                                                                                                                                                                                                                           | `csi-rbdplugin-snapclass`                               |
+| `volumeSnapshotClass.annotations`                          | Specifies the annotations for the rbd volumeSnapshotClass                                                                                                                                                                                                            | `[]`                                                    |
+| `volumeSnapshotClass.clusterID`                            | String representing a Ceph cluster to provision storage from                                                                                                                                                                                                         | `<cluster-ID>`                                          |
+| `volumeSnapshotClass.snapshotNamePrefix`                   | Prefix to use for naming CephFS snapshots                                                                                                                                                                                                                            | `""`                                                    |
```
```suggestion
| `volumeSnapshotClass.snapshotNamePrefix`                   | Prefix to use for naming RBD snapshots                                                                                                                                                                                                                            | `""`                                                    |
```

iPraveenParihar commented on charts/ceph-csi-rbd/README.md:
```diff
@@ -91,143 +91,159 @@ kubectl delete namespace ceph-csi-rbd
+| `volumeGroupSnapshotClass.name`                            | Specifies the rbd VolumeGroupSnapshotClass name                                                                                                                                                                                                                      | `csi-rbdplugin-groupsnapclass`                          |
+| `volumeGroupSnapshotClass.annotations`                     | Specifies the annotations for the rbd volumeGroupSnapshotClass                                                                                                                                                                                                       | `[]`                                                    |
+| `volumeGroupSnapshotClass.clusterID`                       | String representing a Ceph cluster to provision storage from                                                                                                                                                                                                         | `<cluster-ID>`                                          |
+| `volumeGroupSnapshotClass.pool`                            | Ceph pool into which the RBD image shall be created                                                                                                                                                                                                                  | `replicapool`                                           |
```
```suggestion
| `volumeGroupSnapshotClass.pool`                            | The Ceph pool where the RBD group should be created                                                                                                                                                                                                                  | `replicapool`                                           |
```

hswong3i replied:
Similar wording are copy-and-pasted from https://github.com/ceph/ceph-csi/blob/cc407d157ec33e69d8b33652d5232c9187569624/charts/ceph-csi-rbd/README.md?plain=1#L205

iPraveenParihar commented on charts/ceph-csi-rbd/README.md:
```diff
@@ -91,143 +91,159 @@ kubectl delete namespace ceph-csi-rbd
+| `volumeGroupSnapshotClass.annotations`                     | Specifies the annotations for the rbd volumeGroupSnapshotClass                                                                                                                                                                                                       | `[]`                                                    |
+| `volumeGroupSnapshotClass.clusterID`                       | String representing a Ceph cluster to provision storage from                                                                                                                                                                                                         | `<cluster-ID>`                                          |
+| `volumeGroupSnapshotClass.pool`                            | Ceph pool into which the RBD image shall be created                                                                                                                                                                                                                  | `replicapool`                                           |
+| `volumeGroupSnapshotClass.volumeGroupNamePrefix`           | Prefix to use for naming CephFS volumeGroups                                                                                                                                                                                                                         | `""`                                                    |
```
```suggestion
| `volumeGroupSnapshotClass.volumeGroupNamePrefix`           | Prefix to use for naming RBD volumeGroups                                                                                                                                                                                                                         | `""`                                                    |
```

iPraveenParihar commented on charts/ceph-csi-rbd/values.yaml:
```diff
@@ -526,6 +526,73 @@ storageClass:
+  # represent the Ceph cluster in clusterID below
+  clusterID: <cluster-id>
+
+  # Prefix to use for naming CephFS snapshots.
```
```suggestion
  # Prefix to use for naming RBD snapshots.
```

iPraveenParihar replied:
@hswong3i, this is not yet resolved.

black-dragon74 replied:
@hswong3i Please let the reviewer mark the status as resolved (_help keep track of the reviews_)

iPraveenParihar reviewed (dismissed):
Can you please revert the formatting changes from this PR?

hswong3i commented:
> Can you please revert the formatting changes from this PR?

Because the table contents are too complicated for manual formatting, so I am using 'pandoc -f gfm -t gfm --wrap preserve' for format linting.

hswong3i committed:
helm: Support VolumeSnapshotClass and VolumeGroupSnapshotClass

This PR add support for creating VolumeSnapshotClass and
VolumeGroupSnapshotClass with helm charts. Both default with
`create: false` as StorageClass support.

Moreover, it handle the chart release with branch `gh-pages`, by
integrate with <https://github.com/helm/chart-releaser-action>.

The sample helm repo could find from
<https://alvistack.github.io/ceph-ceph-csi/index.yaml>, by:

    helm repo add ceph-csi https://alvistack.github.io/ceph-ceph-csi
    helm repo update
    helm search repo ceph-csi

Signed-off-by: Wong Hoi Sing Edison <hswong3i@pantarei-design.com>

Madhu-1 commented on charts/ceph-csi-cephfs/templates/groupsnapshotclass.yaml:
```diff
@@ -0,0 +1,30 @@
+{{- if .Values.volumeGroupSnapshotClass.groupSnapshotterSecretNamespace }}
+  csi.storage.k8s.io/group-snapshotter-secret-namespace: {{ .Values.volumeGroupSnapshotClass.groupSnapshotterSecretNamespace }}
+{{ else }}
+csi.storage.k8s.io/group-snapshotter-secret-namespace: {{ .Release.Namespace }}
```
This is having indentation problem

Madhu-1 commented on charts/ceph-csi-cephfs/README.md:
```diff
@@ -118,118 +118,134 @@ kubectl delete namespace ceph-csi-cephfs
-| `CSIDriver.seLinuxMount` | Specify for efficient SELinux volume relabeling | `true` |
-| `instanceID`                                   | Unique ID distinguishing this instance of Ceph CSI among other instances, when sharing Ceph clusters across CSI instances for provisioning. | ` ` |
-| `radosNamespaceCephFS`                         | CephFS RadosNamespace used to store CSI specific objects and keys. | ` ` |
+| Parameter                                                  | Description                                                                                                                                                                                                                                                                   | Default                                                 |
```
The whole table is changing, Can you please add it to the end of the table so that we dont need to have complete table change?

hswong3i replied:
IMHO the README.md itself already a bit messy... That would be better if we could tidy up with `pandoc -f gfm -t gfm --wrap preserve`...

But make the PR focus is more important than cleanup something indentation related, so I selectively apply the required changes only.

Madhu-1 commented on charts/ceph-csi-cephfs/templates/groupsnapshotclass.yaml:
```diff
@@ -0,0 +1,30 @@
+{{- if .Values.volumeGroupSnapshotClass.create -}}
+apiVersion: groupsnapshot.storage.k8s.io/v1beta1
```
create it only if the CRD exists not only if its set to true?

hswong3i replied:
Most likely not possible and maybe out of scope, my example: https://github.com/alvistack/ansible-collection-kubernetes/blob/master/playbooks/70-helm_csi_cephfs-install.yml#L71-L90

Since we are not EMBEDDED snapshot-controller into ceph-csi, we couldn't simply check if corresponding CRD already exists by using Helm syntax?

Madhu-1 replied:
its possible with helm capabilities https://v2.helm.sh/docs/chart_template_guide/#built-in-objects we have done this earlier as well  For example https://github.com/ceph/ceph-csi/blob/v3.6.1/charts/ceph-csi-rbd/templates/csidriver-crd.yaml#L1

Madhu-1 replied:
if we do the check we dont need `.create` flag

hswong3i replied:
> its possible with helm capabilities https://v2.helm.sh/docs/chart_template_guide/#built-in-objects we have done this earlier as well For example https://github.com/ceph/ceph-csi/blob/v3.6.1/charts/ceph-csi-rbd/templates/csidriver-crd.yaml#L1

@Madhu-1 may you suggest the corresponding syntax similar as `{{ if semverCompare ">=1.18.0-beta.1" .Capabilities.KubeVersion.Version }}` for me? Since I am just new with Helm Chart templating so not much idea about that :-(

Madhu-1 replied:
`{{- if .Capabilities.APIVersions.Has "groupsnapshot.storage.k8s.io/v1beta1" }}` , we just need to add a check like this

hswong3i replied:
> its possible with helm capabilities https://v2.helm.sh/docs/chart_template_guide/#built-in-objects we have done this earlier as well For example https://github.com/ceph/ceph-csi/blob/v3.6.1/charts/ceph-csi-rbd/templates/csidriver-crd.yaml#L1

I give a check with https://github.com/ceph/ceph-csi/blob/v3.14.0/charts/ceph-csi-rbd/templates/storageclass.yaml#L1 and seems semi-auto with `.create` flag looks much flexible and under control for daily usage?

Or better say: yes, we could use full-auto with `{{- if .Capabilities.APIVersions.Has "groupsnapshot.storage.k8s.io/v1beta1" }}`, but shall we also give similar check for `storageclass.yaml` and remove its `.create`, too?

Madhu-1 replied:
@hswong3i This is special because of the external CRD's where it need to be installed but the storageclasses exists by default., Lets do this only for volumesnapshot and volumegroupsnapshot only

Madhu-1 replied:
@hswong3i did you get sometime to address this if check?

hswong3i replied:
Honestly, I am voting for semi-auto with `.create` flag, rather than full-auto with auto detect:
> Since we are not EMBEDDED snapshot-controller into ceph-csi, we couldn't simply check if corresponding CRD already exists by using Helm syntax?

Moreover, what if we installed snapshot controller BUT setting our `.Values.storageClass.create` as false? I don't really think this as a valid combination...

Madhu-1 replied:
Thats something possible because not everyone wants all the features but we can take it in future PR as well, This is an extra check to ensure the installation doesnt fail by default if the CRD's are not present

hswong3i replied:
> Thats something possible because not everyone wants all the features but we can take it in future PR as well, This is an extra check to ensure the installation doesnt fail by default if the CRD's are not present

Totally agree and that's also how `.create` working: first KISS with true/false options as semi-auto, future PR could upgrade and improve it with full-auto automatic detect and checking.

Madhu-1 commented on charts/ceph-csi-cephfs/values.yaml:
```diff
@@ -375,6 +375,71 @@ storageClass:
+  # Annotations for the Volume Snapshot class
+  # Example:
+  # annotations:
+  #   storageclass.kubernetes.io/is-default-class: "true"
```
Can you please change this annotation for snapshot class

iPraveenParihar replied:
@hswong3i, this is what you need to add for volume snapshot class
```
snapshot.storage.kubernetes.io/is-default-class: "true"
```

Madhu-1 commented on charts/ceph-csi-cephfs/values.yaml:
```diff
@@ -375,6 +375,71 @@ storageClass:
+  # Annotations for the Volume Snapshot class
+  # Example:
+  # annotations:
+  #   storageclass.kubernetes.io/is-default-class: "true"
```
same here

Madhu-1 commented on charts/ceph-csi-cephfs/values.yaml:
```diff
@@ -375,6 +375,71 @@ storageClass:
+  create: false
+  name: csi-cephfsplugin-groupsnapclass
+
+  # Annotations for the Volume Snapshot class
```
annotation for group snapshotclass

Madhu-1 reviewed (commented):
Please do adress the same comments for RBD as well.

Madhu-1 commented on charts/ceph-csi-cephfs/values.yaml:
```diff
@@ -375,6 +375,71 @@ storageClass:
+  # String representing a Ceph cluster to provision storage snapshot from.
+  # Should be unique across all Ceph clusters in use for provisioning,
+  # cannot be greater than 36 bytes in length, and should remain immutable for
+  # the lifetime of the StorageClass in use.
```
volumesnapshotclass in use

Madhu-1 commented on charts/ceph-csi-cephfs/values.yaml:
```diff
@@ -375,6 +375,71 @@ storageClass:
+  # String representing a Ceph cluster to provision storage from.
+  # Should be unique across all Ceph clusters in use for provisioning,
+  # cannot be greater than 36 bytes in length, and should remain immutable for
+  # the lifetime of the StorageClass in use
```
volumegroupsnapshotclass in use

Madhu-1 commented:
@iPraveenParihar PTAL you have requested changes on this PR

Madhu-1 commented:
@mergifyio queue

Madhu-1 commented:
@mergifyio rebase

Madhu-1 commented:
@hswong3i can you please rebase this PR

Madhu-1 commented:
@mergifyio queue

black-dragon74 commented on charts/ceph-csi-rbd/values.yaml:
```diff
@@ -526,6 +526,73 @@ storageClass:
+  # represent the Ceph cluster in clusterID below
+  clusterID: <cluster-id>
+
+  # Prefix to use for naming CephFS snapshots.
```
```suggestion
  # Prefix to use for naming RBD snapshots.
```

hswong3i replied:
DONE

black-dragon74 reviewed (changes requested):
Please address this small nit..

Madhu-1 commented:
@mergifyio queue
